File: services/ledger-service/routers/ledger.py
import uuid

from db import get_db
from fastapi import APIRouter, Depends, Query
from models.org import OrgMembership
from schemas.ledger import LedgerEntryCreate, LedgerEntryResponse, LedgerEntryUpdate, LedgerSummary
from services.customer_service import get_customer
from services.ledger_service import create_ledger_entry, get_ledger_history, update_ledger_entry
from services.security import get_org_member
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(tags=["ledger"])


@router.post("/ledger-entry", response_model=LedgerEntryResponse, status_code=201)
async def create_ledger_entry_endpoint(
    data: LedgerEntryCreate,
    membership: OrgMembership = Depends(get_org_member),
    db: AsyncSession = Depends(get_db),
):
    return await create_ledger_entry(membership.user_id, membership.org_id, data, db)


@router.get("/ledger/{customer_id}", response_model=LedgerSummary)
async def get_ledger_history_endpoint(
    customer_id: uuid.UUID,
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    membership: OrgMembership = Depends(get_org_member),
    db: AsyncSession = Depends(get_db),
):
    customer = await get_customer(customer_id, membership.org_id, db)
    entries, total, total_debit, total_credit, outstanding = await get_ledger_history(
        customer_id, membership.org_id, db, skip=skip, limit=limit
    )
    return LedgerSummary(
        customer_id=customer.id,
        customer_name=customer.name,
        total_debit=str(total_debit),
        total_credit=str(total_credit),
        outstanding_balance=str(outstanding),
        entries=entries,
        total=total,
        skip=skip,
        limit=limit,
    )


@router.put("/ledger-entry/{entry_id}", response_model=LedgerEntryResponse)
async def update_ledger_entry_endpoint(
    entry_id: uuid.UUID,
    data: LedgerEntryUpdate,
    membership: OrgMembership = Depends(get_org_member),
    db: AsyncSession = Depends(get_db),
):
    return await update_ledger_entry(entry_id, membership.org_id, data, db)

File: apps/web-dashboard/lib/api/ledger.ts
import { ledgerClient } from "./client"
import type {
  CustomerCreate,
  CustomerWithBalance,
  LedgerEntryCreate,
  LedgerEntryResponse,
  LedgerSummary,
  PaginatedList,
} from "@/types/api"

// ── Customers ────────────────────────────────────────────────────────────────

export async function getCustomers(
  search?: string,
  skip = 0,
  limit = 20
): Promise<PaginatedList<CustomerWithBalance>> {
  const res = await ledgerClient.get<PaginatedList<CustomerWithBalance>>("/customers", {
    params: { ...(search ? { search } : {}), skip, limit },
  })
  return res.data
}

export async function getCustomer(id: string): Promise<CustomerWithBalance> {
  const res = await ledgerClient.get<CustomerWithBalance>(`/customers/${id}`)
  return res.data
}

export async function createCustomer(data: CustomerCreate): Promise<CustomerWithBalance> {
  const res = await ledgerClient.post<CustomerWithBalance>("/customers", data)
  return res.data
}

// ── Ledger entries ────────────────────────────────────────────────────────────

export async function getLedger(
  customerId: string,
  skip = 0,
  limit = 20
): Promise<LedgerSummary> {
  const res = await ledgerClient.get<LedgerSummary>(`/ledger/${customerId}`, {
    params: { skip, limit },
  })
  return res.data
}

export async function createLedgerEntry(
  data: LedgerEntryCreate
): Promise<LedgerEntryResponse> {
  const res = await ledgerClient.post<LedgerEntryResponse>("/ledger-entry", data)
  return res.data
}

export async function settleEntry(
  id: string,
  is_settled: boolean
): Promise<LedgerEntryResponse> {
  const res = await ledgerClient.put<LedgerEntryResponse>(`/ledger-entry/${id}`, {
    is_settled,
  })
  return res.data
}
